pkg/controller/volume/persistentvolume/pv_controller.go

```diff
@@ -545,6 +551,21 @@ func (ctrl *PersistentVolumeController) syncVolume(volume *v1.PersistentVolume)
 		if err != nil {
 			return err
 		}
+		if !found && metav1.HasAnnotation(volume.ObjectMeta, annBoundByScheduler) {
+			// If PV is pre-bound by scheduler, it's possible on heavy load
+			// that corresponding PVC is not synced in controller yet. So
+			// we double-check PVC in apiserver to make sure we will not
+			// reclaim a PV due to API delay.
+			// Note that only non-released and non-failed volumes will be
+			// updated to Released state when PVC does not eixst.
+			if volume.Status.Phase != v1.VolumeReleased && volume.Status.Phase != v1.VolumeFailed {
+				obj, err = ctrl.kubeClient.CoreV1().PersistentVolumeClaims(volume.Spec.ClaimRef.Namespace).Get(volume.Spec.ClaimRef.Name, metav1.GetOptions{})
```

Comment:
So another theory as to what could be happening here.  PV controller's pvc cache is different from the informer cache.  Whenever a PVC informer event comes in, the pvc gets added to a queue, and the queue is processed by a single claimWorker, which will then add it to the PVC cache.  So if there are many PVC events all at once, then it's possible that the "create pvc event" is stuck in the queue, even though the informer actually saw the update.

So one more thing we could add here is to check the informer cache directly before checking the API server.  That could avoid the extra API call in this scenario, although will not avoid the api call on a normal PVC delete.